src/formatting_stack/util/diff.clj

```diff
@@ -0,0 +1,70 @@
+(ns formatting-stack.util.diff
+  (:require
+   [clojure.java.data :as data]
+   [clojure.java.io :as io]
+   [clojure.spec.alpha :as spec]
+   [clojure.string :as string]
+   [formatting-stack.util :refer [rcomp]]
+   [nedap.speced.def :as speced]
+   [nedap.utils.spec.predicates :refer [present-string?]])
+  (:import
+   (difflib DiffUtils)
+   (io.reflectoring.diffparser.api UnifiedDiffParser)
+   (java.io File)
+   (java.util.regex Pattern)))
+
+(spec/def ::start nat-int?)
+(spec/def ::end nat-int?)
+(spec/def ::filename present-string?)
+
+(speced/def-with-doc ::line-numbers
+  "maps of consecutive removals/changes (as `:start` to `:end`) per `:filename`."
+  (spec/coll-of (spec/keys :req-un [::start ::end ::filename])))
+
+(speced/defn ^::line-numbers diff->line-numbers
+  [^string? diff]
+  (->> (.parse (UnifiedDiffParser.) (io/input-stream (.getBytes diff)))
```

Comment:
> I don't see the importance.

Ensuring that we are performing a valid diff comparison seems better placed as a responsiblity of f-s. Third party libs can be swapped out, have bugs, evolve and lack a precondition system.

It's quite plausible that any of these libs could decide to return an empty diff on invalid input (e.g. nonexisting file, mixed relatve vs. absolute path).

Specs or tests cover us against these in a precise fashion beyond what fine-grained, one-off reasoning can accomplish.

> [...] a regex could be.

No need for regex if that's part of the perceived problem

> A diff does not need to be accessible in order be parsable.

Forbidding non-accessible files rules out plausible bugs

> arbitrary restrictions like these can come as a surprise in the future.

It'd be a good kind of surprise because it would mean that there are corner cases that lack test/spec coverage. It's better than other plausible surprises (e.g. an opaquely wrong result that didn't throw an exception)

Reply:
Still not convinced. This function derives line numbers from a diff. the filepaths are part of the data and do not need to be related to the local filesystem.

If the existence is required, we should put it in the consumer of this fn. 

This function is currently pure (adding filesystem checking would make it impure), and i'd like to keep it that way.

Reply:
> This function is currently pure (adding filesystem checking would make it impure), and i'd like to keep it that way.

Ok, let's leave it as-is